Buffer in Pickler to improve performance.

jjlilley opened the pull request:
Summary:
This change adds a small fixed-size buffer to Pickler to
avoid calling writer_() and the associated downstream checks
on a per-opcode/per-byte basis.

We end up still doing a bounds check in the common case,
but the memcpy() is a fixed size. And we reduce the number
of backend calls.

In practice, this change speeds up the Pickle1MInts benchmark
for me locally from roughly 56msec to 22msec.

Additionally, in this change we convert a few pushIValue() on
typed lists, where we know the type to be double/int/boot to be
pushInt() to bypass a bit of logic.

We should additionally change the Unpickler, though keeping
this separate, since the std::function<> prototype needs to be
changed for this to work (i.e. return size_t rather than bool).

Commit:
Buffer in Pickler to improve performance. (#27720)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/27720

This change adds a small fixed-size buffer to Pickler to
avoid calling writer_() and the associated downstream checks
on a per-opcode/per-byte basis.

We end up still doing a bounds check in the common case,
but the memcpy() is a fixed size. And we reduce the number
of backend calls.

In practice, this change speeds up the Pickle1MInts benchmark
for me locally from roughly 56msec to 22msec.

Additionally, in this change we convert a few pushIValue() on
typed lists, where we know the type to be double/int/boot to be
pushInt() to bypass a bit of logic.

We should additionally change the Unpickler, though keeping
this separate, since the std::function<> prototype needs to be
changed for this to work (i.e. return size_t rather than bool).

Test Plan:
buck test mode/dev-nosan caffe2/test:...
  Benchmark in experimental/jeremyl/c2/SerializationBench.cpp (run in mode/opt)

Differential Revision: D17847174

fbshipit-source-id: 30eace0187f71d9fff13a53013227b780f93c6c0

resistor commented on torch/csrc/jit/pickler.cpp:
```diff
@@ -408,6 +423,7 @@ void Pickler::pushLong(const std::string& data) {
         data.size() > std::numeric_limits<uint32_t>::max(),
         "Cannot pickle a long with a size larger than 4 bytes")
     push<PickleOpCode>(PickleOpCode::LONG4);
+    // TODO: should this be uint32_t instead?
```
Yeah... this looks wrong.  See https://github.com/python/cpython/blob/2.7/Lib/pickletools.py#L677

jjlilley replied:
Thanks. I took this TODO out of the code, and uploading a shortish diff to deal with this.

resistor commented on torch/csrc/jit/pickler.h:
```diff
@@ -183,19 +185,40 @@ class Pickler {
   std::function<void(const char*, size_t)> writer_;
 
+  // Buffer to avoid calling a writer_ on a per-byte basis.
+  std::array<char, 256> buffer_;
```
I think you could simplify a lot of the logic in this patch away by making this a `SmallVector<char, 256>`.  Then you just can just check `size()` rather than having to track `bufferPos_`, etc.

zdevito commented on torch/csrc/jit/pickler.h:
```diff
@@ -183,19 +185,40 @@ class Pickler {
   void push(typename std::common_type<T>::type value) {
     const char* begin = reinterpret_cast<const char*>(&value);
-    writer_(begin, sizeof(T));
+    if (bufferPos_ + sizeof(T) > buffer_.size()) {
```
we should static assert that `sizeof(T) < 256`, otherwise the memcpy is invalid. This seems unlikely to happen, but it is possible if someone changes value to no longer be common_type, and then passes in a large struct.

jjlilley replied:
Good idea. I was half-thinking about this at the time, but it didn't make it into final diff, uploading a new version with this.

zdevito reviewed (approved):
Looks good, thanks!

jjlilley commented:
Would you mind stamping https://github.com/pytorch/pytorch/pull/28043 ?

I used the web-based export-from-phabricator originally with this diff, which alas is suffering from an Infra failure, and apparently using the web tool makes ghexport not work.

The change 28043 is identical except that I addressed the two pending review comments:
  - I added the static assert.
  - I removed the TODO comment re: LONG4 so I could fix that in a followup change

Thanks!
